src/posts/2020-09-19-how-to-write-accessible-forms.md

```diff
@@ -0,0 +1,158 @@
+
+The following are ways to effectively and accessibly communicate form errors:
+- Let the person know there is an error.
+- Let the person access the error, and provide instructions on how to fix it.
+- Allow the person re-submit the form.
+
+
+In the code below, there is an empty [live region](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Live_Regions) which report errors to assistive technologies. Additionally, each field is followed by a _hint_ which is hidden by default but shown if that field does not validate.
+
+```html
+<form>
+  <label for="username">Enter a username</label>
+  <input type="text" id="username">
+  <p class="hint hidden" id="usernameHint" >Your username cannot contain punctuation</p>
+
+  <label for="email">Enter your email address</label>
+  <input type="text" id="email">
+  <p class="hint hidden" id="emailHint">Please enter a valid email address</p>
+
+  <label for="password">Enter a password</label>
+  <input type="password" id="password">
+  <p class="hint hidden" id="passwordHint">Your password must be at least 6 characters</p>
+
+  <div aria-live="assertive" id="message"></div>
+ 
+  <button type="submit">Submit</button>
+</form>
+```
+
+```css
+.hidden{
+  display: none
+}
+
+.hint{
+  color: red;
+}
+```
+If validation errors occur, two things happen:
+- First, the [live region](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Live_Regions) is populated upon form submission, announcing to assistive devices that there are errors.
+
+- Second, each invalid `<input>` is given an [aria-invalid="true"](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_aria-invalid_attribute) attribute and connected to a __hint__ using  an [aria-describedby](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_aria-describedby_attribute) attribute.
+
+Here's the markup for the email field in its invalid state:
+```html
+<label for="email">Your email</label>
+<input type="text" id="email" aria-invalid="true" aria-describedby="emailHint">
+<p class="hint" id="emailHint">Invalid email address.</p>
+```
+
+## Avoid horizontal layout forms unless neccessary
+
+Layout attributes a great deal  to the usability of a site. With CSS, changing a form's visual layout is possible. However, changing the visual layout [create problems](https://youtu.be/8MAvH6vYbDo) for people who rely on the structural layout of the page. As a rule of thumb, your forms should have a meaningful and logical sequence. If possible, avoid making your form inputs next to each other. This is because people have to spend a little more time correlating the labels with their elements, resulting in slower form completion

```

Comment:
Instead of:

>As a rule of thumb, your forms should have a meaningful and logical sequence.

Can we use:

>Because of this, your forms should have a meaningful and logical sequence.

The term "as a rule of thumb" has associations with domestic abuse.